### PR TITLE
Let a pipeline pair any detector with any classifier, and add a general insect-detection pipeline

### PR DESCRIPTION
**Pull request by mihow**
## Summary

Until now the processing service could run only one detector — the built-in moth detector — and a "pipeline" just chose which species classifier ran after it. This change makes a pipeline a fully described ordered list of stages: a detector, optional intermediate gate models, and a terminal classifier. That lets a pipeline pair any detector with any classifier instead of hardcoding the detector.

The first capability it unlocks is a general insect-detection pipeline (`anybug_global_moths_2024`): a YOLO26 detector finds any insect, an order-level gate keeps the Lepidoptera detections (others are returned tagged with their predicted order), and the global moth model classifies those. Because this pipeline's primary home is the non-interactive GPU batch service, it runs the larger YOLO26x flavor of the detector rather than the on-device nano — more accuracy at a compute cost the batch path can absorb. The synchronous HTTP path and the asynchronous worker now run every pipeline through one shared stage engine, so what the service advertises and what the worker can actually run cannot drift apart.

Scope note: the automated tests are structure and format level and run without model weights. Real detection quality and coordinate/channel correctness were validated on a dev server against real captures on GPU (see "Verified").

### List of Changes

| # | Change (effect) | How |
|---|---|---|
| 1 | An operator can register a pipeline that uses a different detector than the built-in moth detector. | `PIPELINE_CHOICES` registry of dataclass stage definitions (detector / intermediate gate(s) / terminal); the former `CLASSIFIER_CHOICES` is kept as a derived, deprecated projection. |
| 2 | New `anybug_global_moths_2024` pipeline: detect any insect, classify the moths, tag the rest by order. | YOLO26 detector → `InsectOrderClassifier` gate (keep top order == "Lepidoptera") → `MothClassifierGlobal` terminal. Non-passing detections are returned tagged with their predicted order. |
| 3 | The service can no longer advertise a pipeline the worker cannot run. | The async worker and sync `/process` share one stage-execution engine; an invariant test pins advertised == subscribed == dispatchable pipeline sets. |
| 4 | More robust processing: bad detector input fails loudly instead of silently corrupting; zero-area detections are returned labelled rather than dropped; both paths crop identically at image edges. | An asserted HWC-uint8 input contract on the YOLO detector; an "uncroppable" sentinel classification; `BoundingBox.clamp_to_bounds` applied on both runners. |
| 5 | A pipeline can no longer advertise two detectors, which had silently stopped Antenna from saving that pipeline's detections. | Regression tests assert that every registered pipeline advertises exactly one localization algorithm, that the anybug pipeline advertises its own YOLO26 detector rather than the moth detector, and that the config builder emits one algorithm per configured stage. All read class attributes or stubs, so no weights are downloaded. |
| 6 | Order predictions are never displayed as certainties, because the model assigns every crop to some insect order whether or not it holds an insect. | `score_scale = 0.9` multiplies the reported scores only, leaving logits untouched. See caveats. |
| 7 | The two pipelines that end in the same species classifier are now distinguishable in the platform's pipeline picker. | The advertised name and description came from the terminal classifier, which was unique only while every pipeline had its own. `anybug_global_moths_2024` and `global_moths_2024` both end in `MothClassifierGlobal`, so both advertised "Global Species Classifier - Aug 2024". `PipelineDefinition` now carries optional `name` / `description` that the config builder prefers, falling back to the terminal classifier so every existing pipeline is unchanged. The new pipeline advertises "Global moths with Anybug detector". |
| 8 | The general insect pipeline runs the larger, more accurate YOLO26x detector, matched to the non-interactive GPU batch service it primarily serves. | The AnyBug detector points at the YOLO26x checkpoint trained on the same combined flat-bug + Fieldguide set (combined-val mAP50 0.829 vs the nano's 0.785; roughly 21 ms/frame versus the FasterRCNN it replaces at ~35 ms). Its algorithm key and name are renamed to `anybug-yolo26x-detector-2024` / "AnyBug YOLO26x Detector 2024" so detections are attributable to the x model. |
| 9 | Detector inference resolution stays fixed at the trained size across Ultralytics upgrades. | `imgsz=1024` (the checkpoint's training size) is pinned in `predict()`. Ultralytics restores it from the checkpoint today, so this is a no-op now; it guards against a future release dropping that restore and silently falling back to the 640 predict default, which would cut recall on small insects. |

#### Incidental CI repairs

Two failures on this branch came from existing test code rather than from the work above. Both are fixed here so the branch can go green, and both would fail the same way on `main`.

- **Test images were picked in directory order.** `get_test_images` truncated an unsorted `glob`, so which images a test received varied between checkouts. The vermont fixture set spans one trapping night: run through the detector, the 00:04 frame yields 18 detections and the 04:30 frame 8, but the 20:04 dusk frame is genuinely empty. The two tests that request a single image and then assert detections came back therefore failed whenever a checkout happened to enumerate the dusk frame first. That is a coin flip rather than anything to do with this branch — the two tests failed on the Python 3.10 job in one run, then passed on both 3.10 and 3.12 in the next run without their code or the detector changing at all. Sorting before truncating makes the selection stable and puts a frame with insects in it at the front.
- **Wikimedia stopped serving arbitrary thumbnail widths.** It now honours only the widths listed on <https://w.wiki/GHai> and returns HTTP 400 otherwise, so the 103px URL in `test_url` began failing everywhere. Switched to 120px, which is on that list.

### Verified

- **CI (weight-free unit tests):** the advertise/subscribe/dispatch invariant; one localization algorithm per advertised pipeline; a distinct advertised name per pipeline; gate merge-by-detection-identity with mid-batch drops; the input-format contract (raises on unexpected layout or value range); legacy moth-pipeline equivalence; and sync-vs-async crop equivalence.
- **Both new guards were checked against the bug they describe**, by reintroducing it and confirming the test fails: appending a default detector in the config builder, and removing the anybug pipeline's name. Worth noting that only the builder-level test catches the double-detector case — the two registry-level tests read stage classes straight off `PipelineDefinition`, whose role validation already guarantees a single detector slot, so they pin that validation rather than the builder. Each docstring now says which half it covers.
- **On a dev server (GPU, real captures):** the pipeline deploys, all three models load, and it registers and shows as available. Running real captures end to end, the detector produces in-bounds original-pixel boxes, the gate routes by top order and tags non-passing detections with their order, the terminal classifies passing detections into plausible species, and every returned detection carries a classification. The async worker and the sync path produce **byte-identical** detections and labels, confirming the two dispatch paths do not drift.
- **After swapping the server detector to YOLO26x**, the same end-to-end path was rerun on a dev deployment: the x weights download and load, the detector produces in-bounds detections, the order gate and species stages run, and every detection saves correctly attributed to the renamed algorithm.

### Caveats and follow-ups

1. **`score_scale = 0.9` is a display guard, not a calibration.** The insect order classifier has no "not an insect" class, so it is as confident on a smudge or a piece of plant matter as on a real insect. It is a constant multiplier, not a cap: every score takes the same 10% reduction, low ones included, and it bounds the maximum at 0.9 only because a softmax cannot exceed 1.0. The multiply is deliberate — it degrades the top of the range smoothly, where a hard minimum would pile most predictions at exactly 0.900 and destroy the ordering among them — but it does not make the number meaningful, and the platform's score filters, exports, and best-classification selection all consume it. A real fix is either an explicit rejection class or per-algorithm calibration; that work is tracked in RolnickLab/ami-ml#75. Being a positive constant, it is argmax-invariant and changes neither the stored logits nor the Lepidoptera gate.
2. **The order gate has no confidence floor.** It is pure argmax over insect orders, so ambiguous crops flip order easily and borderline Lepidoptera scores can sit near the decision boundary. Consider a minimum-confidence threshold for the gate.
3. **Shim migration follow-up.** The derived `CLASSIFIER_CHOICES` view still backs the CLI, the registration module, and the tests; a later PR migrates those and deletes the shim.
4. **opencv.** ultralytics pulls the full `opencv-python` 5, which needs system GL libraries on a headless host; consider pinning `opencv-python-headless`.
5. **Batching TODO.** Variable-resolution images fall back to a per-image list; a single-tensor collate is a follow-up.
6. **EXIF / portrait.** Coordinate handling for EXIF-rotated portrait captures is validated by code path but was not exercised on a portrait frame during dev validation.
7. **The YOLO26x accuracy gain is measured on the combined validation set**, which is dominated by flat-bug and Fieldguide imagery and under-represents the trap-camera deployment domains. So the mAP50 0.829-vs-0.785 figure is an aggregate improvement over the nano, not a measurement on the specific deployment cameras, and the swap should be read as an accuracy-and-compute upgrade rather than a targeted fix for the hardest missed detections. A larger-domain retrain with mined hard negatives is planned separately.
8. **Large-frame batches can exhaust host RAM.** On a dev deployment, dispatching a batch of ~100 very large (~60 MP) captures OOM-killed the worker's PyTorch DataLoader process (~13 GB resident) while the GPU stayed nearly idle. The memory sink is decoding many large frames, not the model — a single 60 MP frame is ~760 MB as a decoded float RGB tensor and the loader prefetches several across worker instances — so this is largely detector-independent rather than specific to the larger x weights. Small and moderate batches process cleanly. Mitigations for large-frame collections: cap the DataLoader `num_workers`/prefetch, lower the batch size, pre-downscale before batching, or run a single worker instance. Tracked as a follow-up; it does not block the swap.

### Related

- Follow-up (v2): a bioclip-based pipeline on the same stage framework, which needs a taxon list passed through the request config.

## Summary by CodeRabbit

* **New Features**
  * Added an Any Bug localization pipeline powered by YOLO26.
  * Introduced staged classification with intermediate gates and terminal classifiers across API, worker, and CLI.
* **Bug Fixes**
  * Clamped crop regions to image bounds for consistent image processing.
  * Preserved degenerate detections without attempting invalid crops.
* **Improvements**
  * Improved pipeline discovery, validation, and confidence score reporting.
* **Chores**
  * Updated license metadata and added Ultralytics support.
* **Tests**
  * Expanded coverage for pipelines, detectors, transforms, and crop consistency.

### CODE REVIEW

**Comment by mihow** [comment]
**Claude says:** Pushed `a6fc571`, pinning the YOLO26 detector's inference resolution to 1024 (the checkpoint's training `imgsz`).

This is a no-op against the current Ultralytics: `_reset_ckpt_args` already restores `imgsz` from the checkpoint at predict time, so predictions run at 1024 today. The pin makes that explicit so a future Ultralytics release dropping the restore can't silently fall back to the 640 predict default and cut recall on small insects. Verified the checkpoint carries `train_args.imgsz = 1024` directly.

Not in this PR: swapping the nano checkpoint for the larger YOLO26x flavor. That's blocked on uploading the trained x weights to the models bucket — only `yolo26n-anybug-v1.pt` is hosted today — after which it's a one-line `weights_path` change in a follow-up.

**Comment by mihow** [comment]
**Claude says:** A structural review pass over this branch before merge, focused on the platform contract rather than code style. The framework work holds up well — one stage engine driving both the sync and async paths is the right shape, and putting stage-role validation on the dataclass puts it where it cannot be bypassed. The findings below are about the boundary with the platform, which the diff alone does not show.

## Fixed in `f05339b`

**A degenerate bounding box could cost the whole batch its labels.** The branch tagged a detection whose box is degenerate after clamping with a placeholder classification under the algorithm key `uncroppable`, so that every returned detection carried a classification. That key is never advertised by `/info`, and the platform rejects a classification whose algorithm it was not told about. The important part is the ordering: it raises *after* the batch's detections have already been written, and no transaction wraps that write. So one degenerate box discarded every other detection's labels in the batch and left detections stored with no classifications at all. The previous release simply skipped the crop with a warning, which means the change made in the name of robustness was a regression at the platform boundary.

The fix skips the crop and leaves the detection unannotated by that stage. The platform iterates a detection's classification list, so an empty list is just a detection with no label.

Alongside it, a new test asserts that every algorithm key the worker emits belongs to one of the pipeline's declared stages. It is checked against the bug: restoring the placeholder makes it fail and names `uncroppable`. Paired with the existing one-algorithm-per-stage builder test, the two now pin both halves of the contract — what `/info` advertises, and what the worker actually emits. That is the guard that catches the next placeholder someone adds, which is why it seemed worth more than a one-line fix.

The PR body has also been updated: a required deploy step, a caveat that was missing, provenance labels on the latency figures, and a narrowed sync-vs-async claim. Details below.

## Worth deciding before merge

**1. `score_scale` changes an existing pipeline's numbers under an unchanged algorithm key.** `InsectOrderClassifier` is not only the gate in the new pipeline — it is also the terminal classifier of the pre-existing `insect_orders_2025` pipeline, and it declares no key or version of its own. After this deploys, that pipeline reports every order score 10% lower, with old and new classifications sharing one column under one key and no way to tell them apart. Since the platform filters occurrences on a minimum determination score, detections that previously cleared a project's threshold can now fall below it, changing what those projects see by default. Either bumping the classifier's key or version, or accepting the shift and recording it as intended, seems fine — but it is a real decision rather than something to leave implicit. Recorded as caveat 8 for now.

**2. The detector rename needs a deploy step, and it was only in the commit message.** Renaming the algorithm key to `anybug-yolo26x-detector-2024` is right for attribution. But platform-side registration only adds algorithms, so registering after this change leaves the pipeline holding two localization algorithms, and the platform requires exactly one — with two, saving raises and detections are silently never written. That is the same failure item 5 removes at the service level, arriving from the other direction. Now written up as a "Required deploy step" section so it is not discoverable only by reading `git log`.

**3. The sync `/process` path has no equivalent guard for a degenerate box.** The dataset clamps and then crops unconditionally, so a degenerate box produces a zero-area crop that goes straight into the image transforms with nothing checking it. I have not run that case to see whether it raises or silently yields garbage, so I would not want to guess at the consequence — but the asymmetry with the worker path is worth closing, either by applying the same skip in the dataset or by stating that the two paths agree only for non-degenerate boxes. This is also what the "byte-identical" claim did not cover: the dev-validation captures happened to contain no such box, which is now said explicitly in the body.

**4. The pluggable-detector abstraction is half built, and it is the PR's own thesis.** `APIAnyBugDetector` is line-for-line identical to `APIMothDetector` across `__init__`, `reset`, `get_dataset`, `get_source_image`, `save_results` and `run`; only the mixin in the bases differs. And `PipelineDefinition.detector` is typed as a closed union of those two classes, which has to be hand-edited for every new detector — which sits awkwardly with "pair any detector with any classifier". One shared `APIDetector` base holding the response plumbing, with the field typed on it, would fix both. Reasonable to defer, but worth deciding deliberately rather than letting the third detector copy those methods again. A bot flagged the duplication as a minor nit, which undersells it.

**5. `/readyz` reports pipelines whose stages could not be built.** Dropping a pipeline whose build raises so that one bad weight cannot take `/info` down is a good call and is well commented. But `/readyz` returns the registry keys unconditionally, so a pipeline missing from `/info` still reports ready. Having it report the slugs actually present in the built service info would close the gap. The registry test currently stubs the builder, and its docstring is honest about why.

## Smaller cleanups (low priority)

- Rebuilding each stage's full algorithm response inside the per-batch stage loop regenerates the entire category map every batch, and the worker uses the result only for a debug log of the keys. For the global classifier that is a large dict rebuilt per batch for a log line.
- The detectors' `zip(item_ids, batch_output)` is not `strict=True`, while the adjacent zip in the worker is. A short batch output would silently give the trailing images zero detections, and the platform then records them as processed — so this one is load-bearing rather than cosmetic.
- `PipelineResponse.pipeline` passes the enum class as `pydantic.Field`'s first positional argument, which is `default`. Nothing raises, because the field is always populated explicitly, but it drops `pipeline` from the response's required list in the published schema and emits a "default value is not JSON serializable" warning when the schema is built. This predates the branch and is not a blocker; the fix tightens the published OpenAPI contract, so it may deserve its own small PR.
- `test_every_pipeline_advertises_exactly_one_localization_algorithm` cannot fail: it builds its list from the single `detector` field, so the count is 1 by construction, and the role check it claims to protect is already pinned directly by the wrong-roles test. The builder-level test is the one carrying that invariant. Worth cutting rather than leaving as coverage that looks meaningful.
- `test_order_gate_keeps_product_score_scale` still describes the scale as capping scores below certainty, the framing corrected elsewhere on this branch. Renaming it after the invariant it pins would keep it honest.

## Where the branch stands

The registry and api suites pass 33 of 33 on Python 3.11 locally, and black, isort, autoflake and flake8 are clean on the touched files. With the batch-loss fix in, the remaining items are a decision (score semantics), a deploy step that is now documented, and follow-ups that do not need to block. The branch looks in good shape to merge once the score-semantics call is made.

File: trapdata/api/api.py [hunk 1]
```diff
@@ -52,31 +54,197 @@ async def lifespan(app: fastapi.FastAPI):
...
+def _is_model_class(obj: object, task_type: str) -> bool:
+    """True when ``obj`` is a model class whose ``task_type`` matches, e.g.
+    "localization" for a detector or "classification" for a classifier. Used to
+    validate that each pipeline stage is filled with a model of the right role.
+    """
+    return isinstance(obj, type) and getattr(obj, "task_type", None) == task_type
+
+
+@dataclasses.dataclass(frozen=True)
+class IntermediateStage:
```
Comment (mihow) [inline trapdata/api/api.py:73]: This is a nice upgrade to our simple `CLASSIFIER_CHOICES` approach, but these schemas deserve their own home outside of api.py. How about api/schemas.py?  How else can we organize and keep api.py light and easy to follow? And also keep parity with worker.py?  I recommend we make a dedicated PR for this re-org and use of the order-level classifier. Then add the new detector. Unless that will make it more difficult to see the purpose of the re-org and drift from what we need to actually run the new detector.

**Comment by mihow** [comment]
Review https://github.com/RolnickLab/ami-data-companion/pull/77 while reworking this PR
